### PR TITLE
[Core][V1][TPU] Enable structured decoding on TPU V1

### PR DESCRIPTION
**Pull request by Chenyaaang**
https://github.com/vllm-project/vllm/issues/11104

Enable structured decoding on V1 TPU. 
The workflow changes from `model forward -> select hidden states -> sample from hidden (compute logits + sampler)` to 
`model forward -> select hidden states -> compute logits -> structured decode -> sampler`

The change has been verified no recompilation.

## pre-compile time
Adding structured deocoding takes around 25 seconds to compile additional 10 xla graphs. 
previous:
<meta charset="utf-8"><div dir="ltr" style="margin-left:0pt;" align="left">
stage | #xla graphs | time(s)
-- | -- | --
backbone | 24 | 103.87
select_hidden_states | 52 | 12.10
sample_from_hidden | 20 | 16.53

</div></b>
now:

<meta charset="utf-8"><div dir="ltr" style="margin-left:0pt;" align="left">
stage | #xla graphs | time(s)
-- | -- | --
backbone | 24 | 103.83
select_hidden_states | 52 | 12.19
compute_logits | 10 | 2.31
**structured_decoding** | **10** | **25.12**
sample_from_logits | 15 | 14.12

</div></b>

### unit test
 `pytest -s tests/v1/entrypoints/llm/test_struct_output_generate.py::test_structured_output`
### benchmark
server: `VLLM_USE_V1=1 VLLM_XLA_CHECK_RECOMPILATION=1 VLLM_USE_V1=1 vllm serve meta-llama/Llama-3.1-8B-Instruct --disable-log-requests --gpu-memory-utilization 0.98 --max-num-batched-tokens 2048 --max-num-seqs 128 --max-model-len 4096 --tensor-parallel-size 1 --no-enable-prefix-caching`
client: `python benchmarks/benchmark_serving_structured_output.py         --backend vllm         --model meta-llama/Llama-3.1-8B-Instruct         --dataset choice         --structured-output-ratio 1.0         --structured-output-backend auto         --request-rate 10         --num-prompts 500`

### CODE REVIEW

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -980,29 +1034,84 @@ def select_hidden_states(self, hidden_states, indices_do_sample):
...
 
+    def prepare_structured_decoding_input(
+        self, logits: torch.Tensor, scheduler_output: "SchedulerOutput"
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        grammar_bitmask = scheduler_output.grammar_bitmask
+        assert grammar_bitmask is not None
+        num_reqs, vocab_size = logits.shape
+
+        # Pad grammar bitmask so that it won't cause recompilation.
+        grammar_bitmask_padded = torch.zeros((num_reqs, vocab_size // 32),
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:1086]: Why it's `vocab_size // 32`, not a `ceil_div`?

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -980,29 +1034,84 @@ def select_hidden_states(self, hidden_states, indices_do_sample):
...
+        # bitmask with the order of requests
+        struct_out_req_batch_indices: dict[str, int] = {}
+        for req_id in self.input_batch.req_ids:
+            mask_index = scheduler_output.structured_output_request_ids.get(
+                req_id)
+            if mask_index is None:
+                continue
+            batch_index = self.input_batch.req_id_to_index[req_id]
+            struct_out_req_batch_indices[req_id] = batch_index
+            grammar_bitmask_padded[batch_index] = torch.from_numpy(
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:1103]: Can we batch the assignment?

**Review by yaochengji — COMMENTED** [review]
Thanks for your contribution!

What's the impact on the execution time when splitting `sample_from_hidden` to two parts?

How does the performance compare with and without structured decoding?

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -886,19 +894,63 @@ def _precompile_select_hidden_states(self) -> None:
...
+            self.compute_logits(dummy_hidden)
+            logger.info("  -- num_seqs: %d", num_reqs)
+        xm.wait_device_ops()
+        end = time.perf_counter()
+        logger.info("Compilation finished in %.2f [secs].", end - start)
+        self._update_num_xla_graphs("compute_logits")
+
+    def _precompile_structured_decoding(self) -> None:
+        logger.info(
+            "Compiling structured_decoding with different input shapes.")
```
Comment (russellb) [inline vllm/v1/worker/tpu_model_runner.py:919]: Do you want to keep this and other logs in this message? They seem more like debug logs to me, so perhaps `logger.debug` if you want to keep them?

File: vllm/v1/worker/tpu_model_runner.py [hunk 4]
```diff
@@ -980,29 +1034,84 @@ def select_hidden_states(self, hidden_states, indices_do_sample):
...
+                          grammar_bitmask: torch.Tensor, logits: torch.Tensor,
+                          arange: torch.Tensor) -> torch.Tensor:
+        return torch.where(
+            require_struct_decoding,
+            self.apply_grammar_bitmask(logits, grammar_bitmask, arange),
+            logits)
+
+    def apply_grammar_bitmask(self, logits, grammar_bitmask, arange):
+        assert (logits.shape[0] == grammar_bitmask.shape[0])
+        logits_cloned = logits.clone()
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:1063]: Thanks for the explanation.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:1086]: > Why it's `vocab_size // 32`, not a `ceil_div`?

yes that's a good catch, it should be ceil_div, I didn't notice it because the vocab_size in qwen and llama3 can both be divided by 32.

Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:1086]: yes that's a good catch, it should be ceil_div, I didn't notice it because the vocab_size in qwen and llama3 can both be divided by 32.

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:1103]: This is the same logics as gpu's. I think the for loop can't be eliminated, but we can batch copy grammar_bitmask to grammar_bitmask_padded, is that what you mean?

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:919]: sg, I can move them to `logger.debug`

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:1103]: Yeah, I think the copy can be batched.

**Comment by Chenyaaang** [comment]
> Thanks for your contribution!
> 
> What's the impact on the execution time when splitting `sample_from_hidden` to two parts?
> 
> How does the performance compare with and without structured decoding?

**Comment by Chenyaaang** [comment]
> What's the impact on the execution time when splitting `sample_from_hidden` to two parts?

The benchmark_serving results are almost the same (qps, latency).
> How does the performance compare with and without structured decoding?

I ran `benchmark_serving_structured_output.py` and `benchmark_serving.py`, both using input-len 1100 and output-len 1, qps is 7.15 vs 8.19, mean e2el is 70s vs 61s

**Comment by yaochengji** [comment]
> I ran `benchmark_serving_structured_output.py` and `benchmark_serving.py`, both using input-len 1100 and output-len 1, qps is 7.15 vs 8.19, mean e2el is 70s vs 61s

I think the performance drop is acceptable. Although there might be some performance optimization chances but we can handle it in the following PRs.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks for the contribution.

Final nit: Per the contributing guide, could you change the title to 
`[Core][V1][TPU] Enable structured decoding`?

**Review by NickLucche — CHANGES_REQUESTED** [review]
Nice work! Left some comments, take a look when you find the time.

File: vllm/v1/worker/tpu_model_runner.py [hunk 5]
```diff
@@ -725,9 +725,17 @@ def execute_model(
...
         tpu_sampling_metadata = TPUSupportedSamplingMetadata.\
             from_input_batch(self.input_batch, padded_num_reqs, self.device)
-        selected_token_ids = self.sample_from_hidden(hidden_states,
+        if scheduler_output.grammar_bitmask is not None:
+            require_struct_decoding, grammar_bitmask_padded = \
+                self.prepare_structured_decoding_input(logits, scheduler_output)
+            logits = self.structured_decode(
+                require_struct_decoding.to(logits.device),
+                grammar_bitmask_padded.to(logits.device), logits,
+                torch.arange(0, 32).to(logits.device))
```
[the remark this answers is not in the document]
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:774]: Done

File: vllm/v1/worker/tpu_model_runner.py [hunk 6]
```diff
@@ -1022,29 +1076,87 @@ def select_hidden_states(self, hidden_states, indices_do_sample):
...
 
+    def prepare_structured_decoding_input(
+        self, logits: torch.Tensor, scheduler_output: "SchedulerOutput"
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        grammar_bitmask = scheduler_output.grammar_bitmask
+        assert grammar_bitmask is not None
+        num_reqs, vocab_size = logits.shape
+
+        # Pad grammar bitmask so that it won't cause recompilation.
+        grammar_bitmask_padded = torch.zeros((num_reqs, cdiv(vocab_size, 32)),
```
[the remark this answers is not in the document]
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py]: Switched to pre-allocated tensors.

**Review by mgoin — APPROVED** [review]
LGTM, great work!
